[Ignatius Oey] iP

ignatiusoey09 opened the pull request:
# Karen
> "Do or do not, there is no try" - Yoda
Karen allows you to keep track of tasks, providing you with peace of mind knowing your tasks are tracked. It's
- Text based
- Easy to learn
- ~~Fast~~ _Super_ fast to use!

All you need to do is:
1. Download it from here
2. Double click it
3. Add your tasks
4. Let Karen manage your tasks for you!

And it is **FREE** 🫨 

Features:
- [ x ] Managing tasks
- [  ] Managing deadlines (Coming soon!)
- [  ] Reminders (Coming soon!)
___
If you are a Java programmer, you can use it to practice Java too. Here is the `main` method
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ignatiusoey09 committed:
renamed Duke to Karen

ignatiusoey09 committed:
Implemented simple text greeting

ignatiusoey09 committed:
Added echo functionality, which repeats user input

ignatiusoey09 committed:
added Add Task & List Task functionality

ignatiusoey09 committed:
list functionality now displays message if no tasks yet

ignatiusoey09 committed:
added Task class file

ignatiusoey09 committed:
Added mark/unmark functionality

ignatiusoey09 committed:
Added subclasses Todo, Deadline and Event which extend Task class

ignatiusoey09 committed:
implemented command parsing for the 3 types of tasks

ignatiusoey09 committed:
added test inputs for automated testing

ignatiusoey09 committed:
Added expected outputs into EXPECTED.TXT

ignatiusoey09 committed:
Edited bat file to work correctly with Karen

ignatiusoey09 committed:
Handled input errors for mark/unmark commands

ignatiusoey09 committed:
handle invalid input: empty todo name

ignatiusoey09 committed:
Handled invalid input: deadlines and events

ignatiusoey09 committed:
Added delete function

ignatiusoey09 committed:
added enum declaration for keywords

ignatiusoey09 committed:
added switch case skeleton for keyword enum

ignatiusoey09 committed:
reformatted main code body from if/else to switch case and enums

ignatiusoey09 committed:
Add Storage class to handle file operations

ignatiusoey09 committed:
Add saveToFile method for Storage class

ignatiusoey09 committed:
Created initial implementation of Storage

ignatiusoey09 committed:
Add abstract method toFileString to Task

ignatiusoey09 committed:
Implement toFileString for Deadline

ignatiusoey09 committed:
Implement toFileString for Event class

ignatiusoey09 committed:
Add getName method in Task class

ignatiusoey09 committed:
Implement toFilesString for Todo class

ignatiusoey09 committed:
Fix issue of toFileString is missing task name

ignatiusoey09 committed:
Adjust Storage.saveToFile to utilize Task.toFileString()

ignatiusoey09 committed:
Add fromFileString method to Task

ignatiusoey09 committed:
Add readFile method to Storage class

ignatiusoey09 committed:
Add JavaDoc comment to Task::fromFileString

ignatiusoey09 committed:
Task: Fix indexing error in fromFileString method

ignatiusoey09 committed:
Storage: Fix relative path for Karen.txt

ignatiusoey09 committed:
Fully implement save to local storage feature

ignatiusoey09 committed:
Merge branch 'branch-Level-7'

Completed the feature for saving task list to local storage. The task
list is saved to ./data/Karen.txt, allowing for Karen to 'remember'
previous interactions.

ignatiusoey09 committed:
Implement date time parsing for Deadline and Event

Changed Deadline and Event class to store their respective datetime
parameters like duedate as java.time.LocalDateTime instead of just a
String

ignatiusoey09 committed:
Storage: Catch errors from invalid datetime format

ignatiusoey09 committed:
Fix parsing error for Event creation in main loop

ignatiusoey09 committed:
Change Deadline and Event datetime output format

Deadline and Event toString() now returns datetimes in
"MMM d yyy hh:mma". Format is from java.time.format.DateTimeFormatter

ignatiusoey09 committed:
Add error message for invalid datetime input

ignatiusoey09 committed:
Merge branch 'branch-Level-8'

Change implementation for Event and Deadline class to store datetimes as
java.time.LocalDateTime instead of String.

ignatiusoey09 committed:
Add TaskList class to store and manage tasks

Extracted task list operations in main loop into TaskList class, in
order to closer align to OOP principles

ignatiusoey09 committed:
Add Ui class

Extracted out print operations in the main loop to Ui class in order
to closer align with OOP principles.

ignatiusoey09 committed:
Ui: Add readInput method

ignatiusoey09 committed:
Storage: Updated JavaDoc for readFile()

ignatiusoey09 committed:
Storage: change readFile to take a TaskList param

Changed the implementation to populate the provided TaskList argument
after reading Karen.txt

ignatiusoey09 committed:
Add Parser class to handle user command parsing

ignatiusoey09 committed:
Add abstract class Command

ignatiusoey09 committed:
Add ExitCommand class, which is executed to end the program

ignatiusoey09 committed:
Add ListCommand, executed to print TaskList

ignatiusoey09 committed:
Add TodoCommand class

ignatiusoey09 committed:
Add DeadlineCommand class

ignatiusoey09 committed:
Add EventCommand class

ignatiusoey09 committed:
Add MarkCommand class

ignatiusoey09 committed:
Add UnmarkCommand class

ignatiusoey09 committed:
Add DeleteCommand class

ignatiusoey09 committed:
Add UnknownCommand class

ignatiusoey09 committed:
Parser: Update parse method

Updated the parse method to manage all possible command cases. Does not
utilise Ui class for now.

ignatiusoey09 committed:
Remove Storage paramter from Command constructors

ignatiusoey09 committed:
TaskList: Add toFileStrings method

toFileStrings returns a List of fileStrings, each fileString is the
save-to-file format its respective Task in the TaskList.

ignatiusoey09 committed:
Storage: Updated saveToFile method

Updated saveToFile to use new TaskList::toFileStrings method.

ignatiusoey09 committed:
TaskList: Add getSize method

ignatiusoey09 committed:
Ui: Add various print methods

Add showMarkMessage, showUnmarkMessage, showAddTaskMessage and
showDeleteMessage methods.

ignatiusoey09 committed:
Complete execute method for mark and unmark

ignatiusoey09 committed:
Add AddTaskCommand class

Consolidated TodoCommand, DeadlineCommand and Event Command into a
single class.

ignatiusoey09 committed:
Delete Todo, Event and Deadline Commands

ignatiusoey09 committed:
Complete DeleteCommand implementation

ignatiusoey09 committed:
Parser: Update parsing to use AddTaskCommand

ignatiusoey09 committed:
Change logic to use 0-indexing

Fixed bug where IndexOutOfBounds error occurs from passing user input
directly to index the List. Solved by decreasing the user input by 1.

ignatiusoey09 committed:
Ui: Add showUnknownInputError method

ignatiusoey09 committed:
UnknownCommand: Update execute method

ignatiusoey09 committed:
Fix Event and Deadline toFileString method

Corrected Event and Deadline not saving datetimes in the correct format
to Karen.txt

ignatiusoey09 committed:
Fix MarkCommand not marking properly

ignatiusoey09 committed:
Ui: Add invalid syntax messages

Added invalid syntax messages for todo, deadline and event commands.

ignatiusoey09 committed:
Parser: Update to use Ui syntax error messages

ignatiusoey09 committed:
Ui: Add more invalid syntax messages

Added invalid syntax messages for mark, unmark and delete commands.

ignatiusoey09 committed:
Parser: Update to use syntax messages from Ui

ignatiusoey09 committed:
Ui: Add showNotANumberError

Added showNotANumberError method, to be used when user is supposed to
enter a digit but enters something else. Also improved showDateTimeError
message to be more descriptive.

ignatiusoey09 committed:
Parser: Update to use Ui error messages

ignatiusoey09 committed:
Adjust dateTimeError message to be clearer

ignatiusoey09 committed:
Reformat to better suit OOP principles

As described by increment A-MoreOOP, extracted several features out into
classes such as Parser, Ui and Storage.

ignatiusoey09 committed:
Merge branch 'branch-A-MoreOOP'

Extracted out features into several classes in order to better align
with OOP design principles.

ignatiusoey09 committed:
Move Task-related classes to main.java.tasks

Moved Task, TaskList, Todo, Deadline and Event classes into
main.java.tasks package.

ignatiusoey09 committed:
Merge branch 'branch-A-Packages'

Arranged files into 3 main packages: util, commands and tasks.

ignatiusoey09 committed:
Merge branch 'branch-A-Gradle'

Started integrating Gradle into the project.

ignatiusoey09 committed:
Refactor code to work with Gradle

Move all files and packages into a main package Karen. Main Karen.java
class must be in a package for 'radlew run' command to work.

ignatiusoey09 committed:
Integrate JUnit for automated unit testing

Added testcases for TaskList and Todo classes.

ignatiusoey09 committed:
Update Gradle shadowjar configs

dominopizzaaaa commented on src/main/java/Karen/main/Karen.java:
Is Karen a package name and supposed to be lowercase? 🤔

dominopizzaaaa commented on src/main/java/Karen/commands/AddTaskCommand.java:
Good job on all the class names! All good!

dominopizzaaaa commented on src/main/java/Karen/util/Storage.java:
Good format for the constants!

dominopizzaaaa commented on src/main/java/Karen/util/Ui.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    public void showWelcome() {
```
Maybe can add a bit more Javadoc for these 3 files in util! 👍

dominopizzaaaa reviewed (commented):
Good job overall! Good organisation too!

ignatiusoey09 committed:
Add JavaDoc comments for clarity

ignatiusoey09 committed:
Refactor code to meet Java coding standard

ignatiusoey09 committed:
TaskList: Add searchTasks method

ignatiusoey09 committed:
TaskList: Add searchTasks method

ignatiusoey09 committed:
Add FindCommand class

ignatiusoey09 committed:
Add FindCommand case to parsing logic

ignatiusoey09 committed:
TaskList: Add fromList method

ignatiusoey09 committed:
Ui: Add messages associated to FindCommand

ignatiusoey09 committed:
Change FindCommand to use Ui messages

ignatiusoey09 committed:
Add message for invalid Find syntax

ignatiusoey09 committed:
Complete implementation for Find command

ignatiusoey09 committed:
Merge branch 'branch-A-JavaDoc'

Provided more JavaDoc comments to improve clarity of code.

ignatiusoey09 committed:
Merge branch 'branch-A-CodingStandard'

Reformat codebase to follow Java coding standard as specifed in CS2103T
course website.

ignatiusoey09 committed:
Merge branch 'branch-Level-9'

Add new feature to search tasks by a substring using the 'find' command.

ignatiusoey09 committed:
Fix packaging issues from merge

ignatiusoey09 committed:
Set to ignore data/Karen.txt

bmanara commented on src/main/java/karen/commands/Command.java:
```diff
@@ -0,0 +1,20 @@
+     * Checks to see if this {@code Command} is of type {@code ExitCommand}
+     * @return boolean
+     */
+    public abstract boolean isExit();
```
Maybe using a verb would be better as a name for this method? 🤔 
`checkExit` might work, since `isExit` feels more like a boolean variable.

bmanara commented on src/main/java/karen/tasks/Todo.java:
```diff
@@ -0,0 +1,18 @@
+        int markedInt = this.isMarked() ? 1 : 0;
+        return String.format("T | %d | %s", markedInt, this.getName());
+    }
+}
```
Great use of naming in tasks! 💯

bmanara commented on src/main/java/karen/util/Ui.java:
```diff
@@ -0,0 +1,204 @@
+                + LINE;
+        System.out.print(output);
+    }
+}
```
Awesome names used for methods in util! 🥇

bmanara reviewed (commented):
LGTM! Just one small nitpick to consider.

juancarlovieri commented on src/main/java/karen/tasks/Task.java:
```diff
@@ -0,0 +1,87 @@
+     * Returns the value of this.isDone
+     *
+     */
+    public boolean isMarked() {
```
I think method names should be in verbs

juancarlovieri commented on src/main/java/karen/tasks/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+     * Returns an array of all the taskStrings for each <code>Task</code> in the <code>TaskList</code>
+     * @return String[]
+     */
+    public String[] toTaskStrings() {
+        if (this.tasks.isEmpty()) {
+            return new String[] {};
+        } else {
+            String[] taskStrings = new String[this.tasks.size()];
+            for (int i = 0; i < this.tasks.size(); i++) {
+                taskStrings[i] = this.tasks.get(i).toString();
+            }
+            return taskStrings;
+        }
+    }
+
+    /**
+     * Returns a {@code List<String>} containing each Task's fileString
+     * @return {@code List<String>}
+     */
+    public List<String> toFileStrings() {
+        if (this.tasks.isEmpty()) {
+            return new ArrayList<>();
+        } else {
+            List<String> fileStrings = new ArrayList<>();
+            for (Task t : tasks) {
+                fileStrings.add(t.toFileString());
+            }
+            return fileStrings;
+        }
+    }
```
Great use of method names! Very intuitive since it follows the `toString` style. However, why is there an inconsistency of return type between the two functions?

juancarlovieri commented on src/test/java/karen/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,32 @@
+public class TaskListTest {
+
+    @Test
+    public void toTaskStrings_emptyTaskList_emptyStringArray() {
```
Are test methods supposed to be in snake case? If so, is it legal to mix between snake and camel cases?

juancarlovieri commented on src/main/java/karen/util/Ui.java:
```diff
@@ -0,0 +1,204 @@
+import java.util.Scanner;
+
+public class Ui {
+    private final static String LINE = "_______________________\n";
```
Good use of screaming snack case in constant names!

juancarlovieri reviewed (commented):
LGTM! Just some minor naming issues. Love the work!

samriddh2145 commented on src/main/java/karen/commands/MarkCommand.java:
```diff
@@ -0,0 +1,27 @@
+public class MarkCommand extends Command{
+    private int index;
+
+    public MarkCommand(int i) {
```
It might be better to use variable names that are clear and descriptive.
``` suggested
public MarkCommand(int index) {
```

samriddh2145 commented on src/main/java/karen/commands/MarkCommand.java:
```diff
@@ -0,0 +1,27 @@
+import karen.util.Storage;
+import karen.util.Ui;
+
+public class MarkCommand extends Command{
```
A space before the opening braces.
```
public class MarkCommand extends Command {
```

samriddh2145 commented on src/main/java/karen/main/Karen.java:
```diff
@@ -0,0 +1,42 @@
+import karen.util.Ui;
+
+public class Karen {
+    private TaskList taskList;
```
Good use of variable names.

samriddh2145 reviewed (commented):
Overall, the code is well-organised and adheres to coding standards.

ignatiusoey09 committed:
Reformat code to conform to checkstyle

ignatiusoey09 committed:
Merge branch 'branch-A-Checkstyle'

ignatiusoey09 committed:
Update build.gradle to use javaFx

ignatiusoey09 committed:
Change Ui methods to return the message printed

ignatiusoey09 committed:
Refactor Command classes

Changed executed method of Command classes to return the Ui response
message as a String. This string is then passed to the GUI later.

ignatiusoey09 committed:
Add image files to represent Karen and user

ignatiusoey09 committed:
Change karen.png

ignatiusoey09 committed:
Karen: Add getKarenResponse method

ignatiusoey09 committed:
Add JavaFX boilerplate

ignatiusoey09 committed:
Add MainWindow class

This class represents the base window for the GUI, including the text
field and the send button.

ignatiusoey09 committed:
Add DialogBox class

This class represents the message sent by each party(user or Karen). It
includes the display picture and the message sent.

ignatiusoey09 committed:
Add fxml for DialogBox and MainWindow classes

ignatiusoey09 committed:
Change build.gradle to use JavaFx

ignatiusoey09 committed:
Add karen.txt to gitignore

ignatiusoey09 committed:
Remove karen.txt

ignatiusoey09 committed:
Merge branch 'branch-Level-10'

Added GUI to Karen.

ignatiusoey09 committed:
Add ShowErrorCommand class

ignatiusoey09 committed:
Adjust Parser.parse method

Changed parse method to return ShowErrorCommand object in case of caught
error.

ignatiusoey09 committed:
Karen: Add assertion to getKarenResponse

Added a assertion that the command from Parse.parse() should never be
null.

ignatiusoey09 committed:
Parser: Fix import statements and code quality

ignatiusoey09 committed:
Reformat code to follow Code Quality standards

ignatiusoey09 committed:
Merge pull request #1 from ignatiusoey09/branch-A-Assertions

Merge branch-A-Assertions into master

ignatiusoey09 committed:
Merge pull request #2 from ignatiusoey09/branch-A-CodeQuality

Merge branch-A-CodeQuality into master

ignatiusoey09 committed:
Add SortCommand class

ignatiusoey09 committed:
Add getter methods for Event and Deadline classes

ignatiusoey09 committed:
TaskList: Add sortByDate method

ignatiusoey09 committed:
Ui: Add messages related to ShowCommand

ignatiusoey09 committed:
Parser: Add recognition for Sort command

ignatiusoey09 committed:
TaskList: Add sortByDate and sortByAlphabet

ignatiusoey09 committed:
Merge branch 'branch-C-sort'

ignatiusoey09 committed:
Add try-catch to catch invalid index errors

ignatiusoey09 committed:
Add header comments

ignatiusoey09 committed:
Add title to GUI window

ignatiusoey09 committed:
Add Ui.png to ip/docs

ignatiusoey09 committed:
Disable horizontal scrolling for scroll pane

ignatiusoey09 committed:
Edit README to include user guide

ignatiusoey09 committed:
Update README to be more detailed
